flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -211,6 +211,34 @@ public void setup() throws IOException {
 
 		BufferPool bufferPool = bufferPoolFactory.get();
 		setBufferPool(bufferPool);
+
+		requestPartitions();
+	}
+
+	private void requestPartitions() throws IOException, InterruptedException {
+		synchronized (requestLock) {
+			if (!requestedPartitionsFlag) {
```

Comment:
Have you considered removing the `requestedPartitionsFlag`. It seems like this was just needed for lazy initialization and can potentially go now.

Reply:
That might be not that simple, since this flag is also used during `updateInputChannel` from unknown channel. I'm not sure, but update probably can happen concurrently to the setup.